### PR TITLE
WIP: Different api authentication strategies for plugin

### PR DESCRIPTION
**Pull request by bdpiparva**
### Pluginsettings

screen shot 2018-02-16 at 5 40 43 pm
screen shot 2018-02-16 at 5 41 12 pm

### Plugin status report
screen shot 2018-02-16 at 5 43 08 pm
#

### Agent status error report
screen shot 2018-02-16 at 5 43 17 pm

### CODE REVIEW

File: src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java [hunk 1]
```diff
@@ -31,31 +34,34 @@ public static KubernetesClientFactory instance() {
...
-        }
+    private KubernetesClient createClientFor(PluginSettings pluginSettings) {
+        LOG.debug(format("Creating config using authentication strategy {0}.", pluginSettings.getAuthenticationStrategy().name()));
+        final ConfigBuilder configBuilder = new ConfigBuilder();
 
-        Config build = configBuilder.build();
-        return new DefaultKubernetesClient(build);
-    }
+        if (pluginSettings.getAuthenticationStrategy() == OAUTH_TOKEN) {
+            configBuilder.withOauthToken(pluginSettings.getOauthToken());
```
Comment (varshavaradarajan) [inline src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java]: Shouldn't the master url be set here too?

File: src/main/java/cd/go/contrib/elasticagent/PluginRequest.java [hunk 2]
```diff
@@ -71,7 +72,7 @@ public Agents listAgents() throws ServerRequestFailedException {
     }
 
     public void disableAgents(Collection<Agent> toBeDisabled) throws ServerRequestFailedException {
-        LOG.debug("[Server Ping] Disabling Agents:"+toBeDisabled.toString());
+        LOG.debug("[Server Ping] Disabling Agents:" + toBeDisabled.toString());
```
Comment (varshavaradarajan) [inline src/main/java/cd/go/contrib/elasticagent/PluginRequest.java]: String.format? That's what we normally use right?

File: src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java [hunk 3]
```diff
@@ -31,31 +34,34 @@ public static KubernetesClientFactory instance() {
...
 
-    public synchronized KubernetesClient kubernetes(PluginSettings pluginSettings) throws Exception {
-        if (pluginSettings.equals(this.pluginSettings) && this.client != null) {
-            return this.client;
+        } else if (pluginSettings.getAuthenticationStrategy() == CLUSTER_CERTS) {
+            configBuilder
+                    .withMasterUrl(pluginSettings.getClusterUrl())
+                    .withCaCertData(pluginSettings.getCaCertData())
+                    .withClientKeyData(pluginSettings.getClientKeyData())
```
Comment (varshavaradarajan) [inline src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java]: Encode the client key?

File: src/main/resources/plugin-settings.template.html [hunk 4]
```diff
@@ -15,94 +15,147 @@
...
+		<legend>Cluster credentials</legend>
+		<div class="row">
+			<div class="columns small-3 medium-2 larger-2 no-padding">
+				<label>Authentication Strategy</label>
+			</div>
+			<div class="columns small-9 medium-10 larger-10 no-padding"
+				 ng-init="authentication_strategy = (authentication_strategy || 'OAUTH_TOKEN')">
+				<input type="radio" ng-model="authentication_strategy" value="OAUTH_TOKEN" id="oauth-token"
+					   ng-checked="(authentication_strategy == 'OAUTH_TOKEN')"/>
+				<label for="oauth-token">OAuth token</label>
```
Comment (varshavaradarajan) [inline src/main/resources/plugin-settings.template.html]: Should we provide a note about how to obtain this for a service account of the user's choice?

File: src/test/java/cd/go/contrib/elasticagent/executors/GetPluginConfigurationExecutorTest.java [hunk 5]
```diff
@@ -46,48 +48,60 @@ public void assertJsonStructure() throws Exception {
...
+                "    \"display-name\": \"Oauth Token\",\n" +
                 "    \"required\": false,\n" +
-                "    \"secure\": false,\n" +
+                "    \"secure\": true,\n" +
                 "    \"display-order\": \"4\"\n" +
                 "  },\n" +
-                "  \"kubernetes_cluster_password\": {\n" +
-                "    \"display-name\": \"Kubernetes Cluster Password\",\n" +
+                "  \"client_key_data\": {\n" +
+                "    \"display-name\": \"Client key data\",\n" +
```
Comment (varshavaradarajan) [inline src/test/java/cd/go/contrib/elasticagent/executors/GetPluginConfigurationExecutorTest.java]: Client key data => Client key

File: src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java (hunk 1 above)
Comment (bdpiparva) [reply src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java]: No we don't need it.

File: src/test/java/cd/go/contrib/elasticagent/executors/GetPluginConfigurationExecutorTest.java (hunk 5 above)
Comment (bdpiparva) [reply src/test/java/cd/go/contrib/elasticagent/executors/GetPluginConfigurationExecutorTest.java]: Following k8s standard, in `~/kube/config` where `certificate-authority` is path to `.crt` file and `certificate-authority-data` is content of `.crt` file.

File: src/main/resources/plugin-settings.template.html (hunk 4 above)
Comment (varshavaradarajan) [reply src/main/resources/plugin-settings.template.html]: Hmm, sure. We should remember to come back and edit the view.

File: src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java (hunk 1 above)
Comment (varshavaradarajan) [reply src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java]: That means that the cluster url field need not be mandatory.

Comment (bdpiparva) [reply src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java]: Yes it not mandatory when authentication strategy is set to `oauth_token`. But mandatory when authentication strategy is set to `cluster_certs`

Comment (varshavaradarajan) [reply src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java]: We saw that cluster ip was mandatory. Please change the UI and the field to make it mandatory.

File: src/main/java/cd/go/contrib/elasticagent/executors/GetPluginConfigurationExecutor.java [hunk 6]
```diff
@@ -30,25 +27,21 @@
...
-    public static final Field KUBERNETES_CLUSTER_PASSWORD = new Field("kubernetes_cluster_password", "Kubernetes Cluster Password", null, false, true, "5");
-    public static final Field KUBERNETES_CLUSTER_CA_CERT = new Field("kubernetes_cluster_ca_cert", "Kubernetes Cluster CA Certificate", null, false, true, "6");
     public static final Map<String, Field> FIELDS = new LinkedHashMap<>();
+    public static final Field GO_SERVER_URL = new GoServerUrlField("go_server_url", "GoCD server URL", false, "0");
+    public static final Field AUTO_REGISTER_TIMEOUT = new PositiveNumberField("auto_register_timeout", "Agent auto-register timeout (in minutes)", "10", false, false, "1");
+    public static final Field MAX_PENDING_PODS = new PositiveNumberField("pending_pods_count", "Maximum pending pods", "10", false, false, "2");
+    public static final Field OAUTH_TOKEN = new NonBlankField("oauth_token", "Oauth token", true, "3");
+    public static final Field CLUSTER_URL = new SecureURLField("kubernetes_cluster_url", "Cluster URL", true, "4");
```
Comment (varshavaradarajan) [inline src/main/java/cd/go/contrib/elasticagent/executors/GetPluginConfigurationExecutor.java]: Cluster URL need not be secure right?  I see that the secure URL field hardcodes secure metadata as false. Rename the class?

File: src/main/java/cd/go/contrib/elasticagent/executors/ShouldAssignWorkRequestExecutor.java [hunk 7]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (varshavaradarajan) [inline src/main/java/cd/go/contrib/elasticagent/executors/ShouldAssignWorkRequestExecutor.java]: Remove this?

File: src/main/java/cd/go/contrib/elasticagent/executors/GetPluginConfigurationExecutor.java (hunk 6 above)
Comment (bdpiparva) [reply src/main/java/cd/go/contrib/elasticagent/executors/GetPluginConfigurationExecutor.java]: Secure means it is `https`. I will rename that class.

**Pull request merged**